rptest: additional iceberg json e2e tests

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
rptest: additional iceberg json e2e tests

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68583-->
<!--43ca43cd-9b70-4a22-bb63-3b57a7956a37-->
<!--31363a92-593f-4027-9e37-a4fd9c7fb654-->
<!--8c67246b-e941-4f55-b308-8ee45e6fc3f2-->
<!--529a31f6-0020-4185-a2f4-8a71918aa7a6-->
<!--e821d5f4-e937-46cc-a8db-471db8304103-->
<!--77f68115-2b85-4e4f-b007-c71c12c1cfdf-->
<!--f9857f08-6eb8-406e-96b7-ae9a785bcbfc-->
<!--14938bac-4d65-4d04-9ea9-9920b7cce824-->
<!--d0ae0474-61c9-43b5-a927-813d07ea5d4f-->
<!--60386def-04fe-46c7-83ce-f64f194c9d67-->
<!--ee166ab1-fc04-4e5b-bc29-68e62302c517-->
<!--3d1031f0-7bf3-4b84-83c2-e73be8658ab0-->
<!--title-->
## Retry command for [Build#68583](https://buildkite.com/redpanda/redpanda/builds/68583)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_json_schema_unicode@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_json_schema_unicode@{"catalog_type":"nessie","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_json_schema_unicode@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_json_schema_unicode@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_json_schema_unicode@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_json_schema_unicode@{"catalog_type":"nessie","cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f217337a-4a42-4512-b6cc-d4baa7f604d1-->
<!--TEST RESULTS-->
<!--d0ac328d-3427-4db2-9991-c5a3f9c5811a-->
<!--TEST RESULTS-->
<!--417c146d-f78c-45c6-b161-83802e459e0a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68583">build#68583</a></summary>


| test_class              | test_method              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------|:-------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"}      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea97-fe7e-4c20-8fab-142a287ae89a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"}      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-5358-4148-bc73-de6421a97a01 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea97-fe7f-45cb-89d6-96e54f6db1fd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-5358-421d-88af-f3ffdb378f22 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea97-fe80-4228-a276-1241d88eb09b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-535a-4c42-af30-da3e120cd9a0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "trino"}      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea97-fe81-4f31-840f-e48a8037bdbd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "trino"}      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-535b-4b22-b5d4-9dcf05874a74 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea97-fe82-469f-9529-47113f5f196e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-535b-438b-9ccd-e8b8df5f0140 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "trino"}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea97-fe82-488e-8073-b4677aaebc0a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeE2ETests        | test_json_schema_unicode | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "trino"}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-535d-46da-9b85-cce83045b0ca | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DisablingPartitionsTest | test_disable             |                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68583#0197ea9d-5358-4148-bc73-de6421a97a01 | FLAKY         | 14/21    | upstream reliability is '89.44174757281553'. current run reliability is '66.66666666666666'. drift is 22.77508 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68636">build#68636</a></summary>


| test_class                   | test_method                             | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:----------------------------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| partition_translator_fixture | test_batching                           |                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/68636#0197ef1c-e9f2-4f8b-a8c3-26aeaa6677bb | FAIL          | 0/1      |                                                                                                                                                                         |
| ClusterConfigTest            | test_valid_settings                     |                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68636#0197ef3a-7110-479e-b474-74bb6c2b0f7b | FLAKY         | 19/21    | upstream reliability is '83.65287588294652'. current run reliability is '90.47619047619048'. drift is -6.82331 and the allowed drift is set to 50. The test should PASS |
| DeleteRecordsTest            | test_delete_records_topic_policy_change | {"cloud_storage_enabled": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68636#0197ef3a-710e-4d37-966d-a61e13d6aa80 | FLAKY         | 20/21    | upstream reliability is '99.71428571428571'. current run reliability is '95.23809523809523'. drift is 4.47619 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68729">build#68729</a></summary>


| test_class                | test_method             | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudArchiveRetentionTest | test_delete             | {"cloud_storage_type": 1, "retention_type": "retention.ms"}                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68729#0197f48a-64b8-4756-a490-2b2402e328c3 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| CloudStorageChunkReadTest | test_read_chunks        |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68729#0197f48c-4a0a-4434-84f2-c638f4801179 | FLAKY         | 20/21    | upstream reliability is '98.4375'. current run reliability is '95.23809523809523'. drift is 3.1994 and the allowed drift is set to 50. The test should PASS            |
| MaintenanceTest           | test_maintenance_sticky | {"use_rpk": false}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68729#0197f48c-4a0e-4b38-a73f-e42b9798576f | FLAKY         | 19/21    | upstream reliability is '99.0074441687345'. current run reliability is '90.47619047619048'. drift is 8.53125 and the allowed drift is set to 50. The test should PASS  |
| RaftAvailabilityTest      | test_one_node_down      |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68729#0197f48a-64b6-4c84-89d1-d39d28e66996 | FLAKY         | 20/21    | upstream reliability is '97.52808988764045'. current run reliability is '95.23809523809523'. drift is 2.28999 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations    | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68729#0197f48a-64b4-4259-974a-803ad537f80c | FLAKY         | 20/21    | upstream reliability is '99.11894273127754'. current run reliability is '95.23809523809523'. drift is 3.88085 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

nvartolomei committed:
iceberg: avoid throwing exceptions from json conversion

The caller treats exceptions as really exceptional cases and do not even
redirect records to DLQ so we need to convert the code to avoid throwing
or place a "firewall" to catch exceptions and return them as values.

Chose to do the later for brevity.

nvartolomei committed:
iceberg: failing to transalte a value is at most a warn

These failures are caused by user inputs/configuration and do not
indicate a bug in redpanda code so use the warn log level.

I'm introducing some e2e tests trigger this code path and tests fail
with bad log lines otherwise.

I'm choosing not to ignore the error because I expect users to run into
this case quite frequently too. Could be a debug log too.

nvartolomei committed:
datalake: unsupported json schema feature is a warn log

Logging with error level is a good default because in general we should
be able to build an intermediate representation (ir) from all the
schemas that were also accepted by the schema registry.

However, we intentionally do not support all json schemas, at least for
now. For schemas that we don't support we throw a nice exceptions which
helps differentiate unsupported schema vs unexpected schema.

To avoid error log noise and bad log lines in tests we log a warning for
such schemas instead of errors.

nvartolomei committed:
test_utils: coro versions for STR{EQ,NE} assertions

nvartolomei commented:
Ops. Had to fix some bugs @wdberkeley. Please re-review. 🙇‍♂️

andrwng commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -276,4 +277,18 @@ deserialize_json(iobuf buf, const json_conversion_ir& ir) {
+ss::future<value_outcome>
+deserialize_json(iobuf buf, const json_conversion_ir& ir) {
+    try {
+        // Firewall for exceptions during deserialization because the caller
+        // does not expect exceptions to be thrown.
```
👍

andrwng commented on src/v/datalake/record_schema_resolver.cc:
```diff
@@ -132,6 +132,12 @@ checked<resolved_type, type_resolver::errc> translate_json_schema(
           .id = {.schema_id = id, .protobuf_offsets = std::nullopt},
           .type = std::move(root_struct.value())};
+    } catch (iceberg::conversion::json_schema::unsupported_feature_error& e) {
+        vlog(
+          datalake_log.warn,
+          "JSON schema translation failed due to unsupported feature: {}",
+          e.what());
+        return type_resolver::errc::translation_error;
```
>Exceptions in this place cause the entire translation pipeline to stop.
>Records don't even go to DLQ in that case.

I'm confused by this -- it doesn't look like this commit changes whether things end up in the DLQ or not, given it's still returning `translation_error`. The commit message goes on to say that the records do end up in the DLQ. Can you help me understand the case where records don't go to the DLQ?

nvartolomei replied:
🤦‍♂️ updated the commit message as it did not make any sense.

andrwng reviewed (approved):
Changes LGTM, but I'm confused about one of the commit

dotnwat commented on src/v/datalake/record_schema_resolver.cc:
```diff
@@ -132,6 +132,12 @@ checked<resolved_type, type_resolver::errc> translate_json_schema(
               std::move(iceberg_ir.value()))),
           .id = {.schema_id = id, .protobuf_offsets = std::nullopt},
           .type = std::move(root_struct.value())};
+    } catch (iceberg::conversion::json_schema::unsupported_feature_error& e) {
```
nit: `const&`

dotnwat commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -276,4 +277,18 @@ deserialize_json(iobuf buf, const json_conversion_ir& ir) {
+        co_return co_await deserialize_json_impl(std::move(buf), ir);
+    } catch (const value_conversion_exception& e) {
+        co_return e;
+    } catch (std::exception& e) {
```
nit: `const&`
